service/src/main/java/io/camunda/service/UserTaskServices.java

```diff
@@ -40,11 +58,176 @@ public UserTaskServices withAuthentication(final Authentication authentication)
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey = new HashMap<>();
+      final VariableServices variableServices =
+          new VariableServices(searchClient, transformers, authentication);
+      boolean firstIteration = true;
+
+      for (final VariableValueFilter variableValueFilter : query.filter().variableFilters()) {
+        processVariableFilter(
+            variableValueFilter, variableServices, scopeKeyMapByProcessInstanceKey, firstIteration);
+        firstIteration = false;
+      }
+
+      // Collect all unique processInstanceIds and scopeKeys
+      final List<Long> allProcessInstanceIds =
+          new ArrayList<>(scopeKeyMapByProcessInstanceKey.keySet());
+      final List<Long> allScopeKeys =
+          scopeKeyMapByProcessInstanceKey.values().stream()
+              .flatMap(set -> Optional.ofNullable(set).orElse(Collections.emptySet()).stream())
+              .filter(Objects::nonNull)
+              .collect(Collectors.toList());
+
+      if (allProcessInstanceIds.isEmpty() && allScopeKeys.isEmpty()) {
+        return new SearchQueryResult<>(0, Collections.emptyList(), null);
+      }
+
+      final UserTaskFilter mergedFilter =
+          getUserTaskFilter(query, allProcessInstanceIds, allScopeKeys);
+
+      // Create a new UserTaskQuery with the merged filter
+      final UserTaskQuery modifiedQuery =
+          UserTaskQuery.of(b -> b.filter(mergedFilter).sort(query.sort()).page(query.page()));
+
+      return executor.search(modifiedQuery, UserTaskEntity.class);
+    }
     return executor.search(query, UserTaskEntity.class);
   }
 
+  private static UserTaskFilter getUserTaskFilter(
+      final UserTaskQuery query,
+      final List<Long> allProcessInstanceIds,
+      final List<Long> allScopeKeys) {
+    final UserTaskFilter originalFilter = query.filter();
+
+    // Create a new UserTaskFilter with the merged keys and existing values
+    return new UserTaskFilter(
+        originalFilter.userTaskKeys(),
+        originalFilter.userTaskDefinitionIds(),
+        originalFilter.processNames(),
+        originalFilter.assignees(),
+        originalFilter.userTaskState(),
+        originalFilter.processInstanceKeys() != null
+                && originalFilter.processInstanceKeys().isEmpty()
+            ? allProcessInstanceIds
+            : originalFilter.processInstanceKeys(),
+        originalFilter.processDefinitionKeys(),
+        allScopeKeys,
+        originalFilter.candidateUsers(),
+        originalFilter.candidateGroups(),
+        originalFilter.created(),
+        originalFilter.completed(),
+        originalFilter.canceled(),
+        originalFilter.creationDateFilter(),
+        originalFilter.completionDateFilter(),
+        originalFilter.dueDateFilter(),
+        originalFilter.followUpDateFilter(),
+        originalFilter.variableFilters(),
+        originalFilter.tenantIds());
+  }
```

Comment:
:x: This will cause issues and must always be adjusted whenever the filter gets extended or whatever. Instead, let's try to move it to the filter and introduce a first concept of allowing to "extend/adjust" a filter, for example:

```java
public final record UserTaskFilter(
  ...) implements FilterBase {

  public static Builder extending(final UserTaskFilter filterToExtend) {
    return new Builder(filterToExtend);
  }

  public static final class Builder implements ObjectBuilder<UserTaskFilter> {


    public Builder() {}
    private Builder(final UserTaskFilter filter) {
      this.userTaskKeys = filter.userTaskKeys();
      ...;
    }
  }
}
```

That way, at this place, it could do the following

```java
final var newFilter = UserTaskFilter.extending(query.filter()).processInstanceKey(allProcessInstanceIds).build();
return UserTaskQuery.of(b -> b.filter(newFilter).sort(query.sort()).page(query.page()));
```